genie-biogem/src/fortran/biogem_data.f90

```diff
@@ -971,54 +971,93 @@ SUBROUTINE sub_data_update_tracerrelationships()
+       conv_ocn_sed(is_POP,io_ALK) = 0.0
+       conv_sed_ocn(io_ALK,is_POC) = 0.0
+       conv_ocn_sed(is_POC,io_ALK) = 0.0
+    elseif (ocn_select(io_NO3)) then
+       conv_sed_ocn(io_ALK,is_PON) = -conv_sed_ocn(io_NO3,is_PON)
+       conv_ocn_sed(is_PON,io_ALK) = 1.0/conv_sed_ocn(io_ALK,is_PON)
+       conv_sed_ocn(io_ALK,is_POP) = 0.0
+       conv_ocn_sed(is_POP,io_ALK) = 0.0
+       conv_sed_ocn(io_ALK,is_POC) = 0.0
+       conv_ocn_sed(is_POC,io_ALK) = 0.0
+    else
+       conv_sed_ocn(io_ALK,is_PON) = 0.0
+       conv_ocn_sed(is_PON,io_ALK) = 0.0
+       if (abs(par_bio_red_POP_POC*par_bio_red_POP_PON*par_bio_red_PON_ALK) > const_real_nullsmall) then
           if (ctrl_bio_red_ALKwithPOC) then
              conv_sed_ocn(io_ALK,is_POC) = (1.0/par_bio_red_POP_POC)*par_bio_red_POP_PON*par_bio_red_PON_ALK
              conv_ocn_sed(is_POC,io_ALK) = 1.0/conv_sed_ocn(io_ALK,is_POC)
+             conv_sed_ocn(io_ALK,is_POP) = 0.0
+             conv_ocn_sed(is_POP,io_ALK) = 0.0
           else
+             conv_sed_ocn(io_ALK,is_POC) = 0.0
+             conv_ocn_sed(is_POC,io_ALK) = 0.0
              conv_sed_ocn(io_ALK,is_POP) = par_bio_red_POP_PON*par_bio_red_PON_ALK
              conv_ocn_sed(is_POP,io_ALK) = 1.0/conv_sed_ocn(io_ALK,is_POP)
           end if
+       else
+          conv_sed_ocn(io_ALK,is_POC) = 0.0
+          conv_ocn_sed(is_POC,io_ALK) = 0.0
+          conv_sed_ocn(io_ALK,is_POP) = 0.0
+          conv_ocn_sed(is_POP,io_ALK) = 0.0
        end if
-    else
-       conv_sed_ocn(io_ALK,is_PON) = 0.0
-       conv_ocn_sed(is_PON,io_ALK) = 0.0
-       conv_sed_ocn(io_ALK,is_POP) = 0.0
-       conv_ocn_sed(is_POP,io_ALK) = 0.0
     end if
+    ! O2 (of P, N, C)
     ! update O2 demand associated with organic matter (taken as the carbon component)
     ! reduce O2 demand associated with C (and H) oxidation => treat N and P explicitly
     ! NOTE: set no PON O2 demand if NO3 tracer not selected (and increase POC O2 demand)
     ! NOTE: NO3 uptake assumed as: 2H+ + 2NO3- -> 2PON + (5/2)O2 + H2O
     !       (and as implemented, ber N, this ends up as (5/2)/2 = 5.0/4.0
+    ! NOTE: to balance the uptake of NO3 into organic matter
+    !       [2H+ + 2NO3- -> 2PON + (5/2)O2 + H2O]
+    !       with the release of N as ammonium and subsequent oxidation to NO3 ...
+    !       [NH4+ + 2O2 -> NO3- + 2H+ + H2O]
+    !       the remin of PON needs to be adjusted in order that everything is conserved:
+    !       2PON + 3H2O + 2H+ --> 2NH4+ + (3/2)O2
+    !       and per N:
+    !       PON + (3/2)H2O + H+ --> NH4+ + (3/4)O2
+    if (ctrl_bio_red_O2withPOC) then
+       conv_sed_ocn(io_O2,is_POP) = 0.0
+       conv_ocn_sed(is_POP,io_O2) = 0.0
+    else
+       conv_sed_ocn(io_O2,is_POP) = -4.0/2.0
+       conv_ocn_sed(is_POP,io_O2) = 1.0/conv_sed_ocn(io_O2,is_POP)
+    endif
+    if (ocn_select(io_NH4)) then
+       conv_sed_ocn(io_O2,is_PON) = (3.0/4.0)
+       conv_ocn_sed(is_PON,io_O2) = -(4.0/5.0)
```

Comment:
What to do about phytoplankton uptake of NH4? conv_ocn_sed(is_PON,io_O2) should be zero as NH4+ -> PON + 4H+